### ISSUE TITLE
Make #each use _id in object or as getter in object's prototype

### ISSUE THREAD

**Issue by lassombra** [meteor/meteor#6124]
When feeding #each an array of objects, if the objects have _id in them at the instance level than each uses the full range of observable related dom transformations which means that interaction with third party libraries doesn't break, because each keeps track of the actual dom nodes that move and when, say, sortable moves them, it respects that they are already moved.

However if _id is a getter up the objects prototype chain, each does not use that and falls back to index based "change" notification which can easily break in this situation.  Making #each more aggressively use _id would be a better solution in many cases I think.

in packages/observe-sequence/observe_sequence.js:
```
seqChangedToArray = function (lastSeqArray, array, callbacks) {
  var idsUsed = {};
  var seqArray = _.map(array, function (item, index) {
    var id;
    if (typeof item === 'string') {
      // ensure not empty, since other layers (eg DomRange) assume this as well
      id = "-" + item;
    } else if (typeof item === 'number' ||
               typeof item === 'boolean' ||
               item === undefined) {
      id = item;
    } else if (typeof item === 'object') {
      id = (item && _.has(item, '_id')) ? item._id : index;
    } else {
      throw new Error("{{#each}} doesn't support arrays with " +
                      "elements of type " + typeof item);
    }

    var idString = idStringify(id);
    if (idsUsed[idString]) {
      if (typeof item === 'object' && '_id' in item)
        warn("duplicate id " + id + " in", array);
      id = Random.id();
    } else {
      idsUsed[idString] = true;
    }

    return { _id: id, item: item };
  });

  return seqArray;
};
```
The line `id = (item && _.has(item, '_id')) ? item._id : index;`
could be changed to `id = (item && 'id' in item) ? item._id : index;`

In this case objects which hide the _id in a prototype chained getter (such as an ORM type implementation of denormalized documents, just as an example of my use case) would still work.

**Comment by avital** [comment]
Dupe of #4946. I think we should just go ahead and do it.

**Comment by lassombra** [comment]
It's not exactly a dupe.  Checking in two different files.  This one specifically affects how observer operates on objects and handles _in where #4946 is modifying minimongo.  Specifically #4946 won't fix blaze each, while this will.

**Comment by avital** [comment]
@lassombra Yeah you're right. I didn't look carefully enough. We should fix both. (But using `"_id" in foo`, not `foo._id` which would incorrectly match an _id of 0)

**Comment by dgreensp** [comment]
Sounds good.

To keep things straight, here are the three potential ways of checking for `_id`:
1. `foo._id` - gives a false negative on falsy `_id`, but otherwise ok
2. `_.has(foo, "_id")` - doesn't use the prototype chain :(
3. `"_id" in foo` - this is the best one :)

Ideally, we should grep package code for occurrences of (1) and (2) and change them to (3) if it matters, keeping in mind that in many (most? all?) cases it doesn't matter because the object being tested for an `_id` property just came out of Mongo, or is part of a query and not a document at all.  However, occurrences of (1) are not so bad because falsy _ids [don't work anyway](https://github.com/meteor/meteor/pull/4946#issuecomment-136868070).

**Comment by lassombra** [comment]
That is correct, falsey _id breaks things all over.  I think 1 is fine, 3 is technically correct.

I can confirm that my PR #6125 solves the issue that I originally posted about however, as I'm now using it in qa without issue.

In my implementation, I am hiding the mongo object behind a data management object (ORM like) that uses accessors to fetch the data, and create fine grained dependencies on just the data you actually use (basically automatically trimming "fields" down in your mongo query) as well as to handle automatic saving of changes.  Updates are saved in near real time and also travel to a separate API server.  So that is why I have wrapped _id in a getter, which #each was missing.  When I implemented this change locally and even when put to qa with this single line change, each is handling it perfectly.
